network/src/peer/mod.rs

```diff
@@ -195,120 +194,76 @@ where
-                        maybe_request = self.requests_rx.next() => {
-                            if let Some(request) = maybe_request {
-                                self.handle_outbound_request(request, &mut write_reqs_tx).await;
-                            } else {
-                                // This branch will only be taken if all PeerRequest senders for this Peer
-                                // get dropped.
-                                break;
-                            }
-                        },
-                        maybe_message = reader.next() => {
-                            match maybe_message {
-                                Some(message) =>  {
-                                    if let Err(err) = self.handle_inbound_message(message, &mut write_reqs_tx).await {
-                                        warn!(
-                                            NetworkSchema::new(&self.network_context)
-                                                .connection_metadata(&self.connection_metadata),
-                                            error = %err,
-                                            "{} Error in handling inbound message from peer: {}, error: {}",
-                                            self.network_context,
-                                            remote_peer_id.short_str(),
-                                            err
-                                        );
-                                    }
-                                },
-                                None => {
-                                    info!(
-                                        NetworkSchema::new(&self.network_context)
-                                            .connection_metadata(&self.connection_metadata),
-                                        "{} Received connection closed event for peer: {}",
-                                        self.network_context,
-                                        remote_peer_id.short_str()
-                                    );
-                                    self.close_connection(DisconnectReason::ConnectionLost);
-                                }
-                            }
-                        },
-                        // Drive the queue of pending inbound rpcs. When one is
-                        // fulfilled by an upstream protocol, send the response
-                        // to the remote peer.
-                        maybe_response = self.inbound_rpcs.next_completed_response() => {
-                            if let Err(err) = self.inbound_rpcs.send_outbound_response(&mut write_reqs_tx, maybe_response).await {
+        let reason = loop {
+            if let State::ShuttingDown(reason) = self.state {
+                break reason;
+            }
+
+            futures::select! {
+                // Handle a new outbound request from the PeerManager.
+                maybe_request = self.peer_reqs_rx.next() => {
```

Comment:
This isn't particular to this PR, but I like to call these `futures::select!` blocks we have nesting hell.  Since we have so much nesting it goes on and on and on.

After all these refactors, we should probably come up with some standardization on how to generally do these, e.g. each block is a function, or whatever.

Reply:
definitely. I wonder if we could take it even further and enforce a stricter separation between the outer event loop and the inner state machine. Maybe this is too much boilerplate, but you could imagine an inner state machine that only exposes like `handle_{event}` methods and then the outer event loop that just looks like

```rust
futures::select! {
  event = async_event_source => self.inner.handle_<...>(event),
  // ..
}
```

it would probably make it easier to test the inner state machine without forcing an implicit dependency on the order that the event loop processes its events.

would also make it easier to fuzz the state machine; you could sample a list of events and feed them into the state machine directly, then make some assertions about the end state.

the big caveat here is not testing a single state machine in isolation, but it doesn't really support testing many interacting state machines. we would need to mock out the event "transport" between state machines more effectively somehow.

Reply:
Let's keep it separate scope of this PR, but its definitely something I want to tackle on a one off day.  Just to at least clean it up so it's always of that form.  Especially since the formatter really doesn't like dealing with whatever is inside the `futures::select!` since it's a macro.